[Liao Keng I] ip

KengXIII opened the pull request:
# Duke

> "2 🐝 or 🚫 2 🐝" - William Shakespeare

Welcome to my Duke repository! It is simple, but definitely has all you need in a task manager.

```
Fairly Simplistic
```

### Introduction
DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

### How to use?
1. All you need to do is,

* download it from here.
* double-click it.
* add your tasks.
* let it manage your tasks for you 😉

2. That's it!

All for absolutely free!

### Features

- [x] Managing tasks
- [ ] Managing deadlines (<em>to be implemented</em>)
- [ ] Reminders (<em>to be implemented</em>)
- [ ] Help function using `/help`

[Click here to find out more!](https://github.com/KengXIII/ip)
Stay tuned for more updates in the future!

Commit:
Add Gradle support

KengXIII committed:
Running the code now greets users and echo user's input. Programme stops when 'bye' is input

KengXIII committed:
Added the ability to store whatever text entered by the user and display them back to the user when requested.

KengXIII committed:
objects added to the array are now of class Task

KengXIII committed:
added the markIsDone method. users can now mark todos as complete

KengXIII committed:
Add support for tracking three types of tasks: todo, deadline & event

KengXIII committed:
added the input and expected test results

KengXIII committed:
Duke now throws NullDescription and InvalidCommand exception

KengXIII committed:
change array of String to ArrayList<Task>

KengXIII committed:
Add saveText function that writes Task to local

KengXIII committed:
Update constructor of deadline, event

KengXIII committed:
Add loadText function that reads Task from local

KengXIII committed:
Change constructor to take in 'completed' boolean

KengXIII committed:
Add localdate and localtime class for dates

KengXIII committed:
Change dates to localdate and localtime

KengXIII committed:
Fix constructors of deadline and event to use a standard format

KengXIII committed:
Merge branch 'branch-level-7'

KengXIII committed:
Merge branch 'branch-level-8'

KengXIII committed:
Add new Storage, Ui and TaskList class

KengXIII committed:
Add the command classes Done,Delete,Add,Exit

KengXIII committed:
Fix printf issues in all classes from println to printf

KengXIII committed:
Change toString exception methods to getMessage

KengXIII committed:
Add method to identify empty description

KengXIII committed:
Update AddCommand to now detect for empty description and invalid date format

KengXIII committed:
Refactor codes into packages - command, exception, storage, task, ui

KengXIII committed:
Create JUnit test within test folder

KengXIII committed:
Package the app into jar file

KengXIII committed:
Add new ListCommand that shows the list of task from local storage

KengXIII committed:
Fix the exception thrown by DeleteCommand and DoneCommand to IndexOutOfBoundsException

KengXIII committed:
Change string comparison in read method of Storage from == to .equals

KengXIII committed:
Write Javadocs for Duke class

KengXIII committed:
Add JavaDocs to all clases except Ui

KengXIII committed:
Change code in command to follow camel-casing

KengXIII committed:
Merge branch 'branch-A-CodingStandard'

KengXIII committed:
Add fully functional find method

KengXIII committed:
Merge branch 'branch-Level-9'

TohWeiJie01 commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,82 @@
+     * @throws InvalidDateFormat If AddCommand is intended but the date format is not followed
+     */
+    public static Command parse(String input) throws InvalidCommand, NullDescription, InvalidDateFormat {
+        String[] output = input.split(" ");
```
Should the name of the array be outputs?

KengXIII replied:
I originally named it output as it processed the string into individual words. On second thought, I think it might be ambiguous with the new Command that it returns. Thanks for the suggestion!

TohWeiJie01 reviewed (commented):
LGTM. The naming conventions for your code generally seems good to me. Good job 😄!

xiongjya commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,108 @@
+
+    private void verifyDate(String dateInfo, String type) throws InvalidDateFormat {
+        switch(type) {
+            case "deadline":
```
In order to better abide by the Java coding standard, that states that there is no indentation for case clauses, would it be better to remove the indentation here? (and similarly to other areas in your code if you have used switch, case)

Timothyoung97 replied:
In general, I think the `addCommand.java` is pretty well done abiding by the coding standard mostly.  However, I believe for `switch` cases in this module are supposed to be aligned with the `switch`? (Same as Jingya's comment)

KengXIII replied:
Thanks for the catch, I'll amend them in the future commits!

xiongjya commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    public static Command parse(String input) throws InvalidCommand, NullDescription, InvalidDateFormat {
+        String[] output = input.split(" ");
+        if (output.length == 0 || output[0].isEmpty() || output[0].equals(" ")) {
```
It seems that the three conditions are rather similar, do you think that you can abstract at least one of them away (so it would be easier for others to read/be more straight to the point)?

KengXIII replied:
I have removed the 2nd case as I cannot think of an example that turns the statement true 🤔 I will keep that in mind and figure out why I had that before, thanks

xiongjya commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,41 @@
+        try {
+            Task deletedTask = tasks.deleteTask(taskNumber);
+            storage.write(tasks.getTaskList());
+            System.out.printf("\tNoted. I've removed this task:\n" +
```
The coding standard suggests to break before an operator, in this case '+', so would it be better if this minor detail is changed to follow the standard?

xiongjya reviewed (commented):
LGTM 👍 The use of abstract classes for commands and tasks is good, and it is great that you have considered and implemented multiple types of errors/exceptions, but do keep in mind that you have a manageable number of classes/files to work on!

The rest that are suggested would simply be cosmetic changes, as your code looks pretty short (not over 30 lines for function declarations other than switch statements) and avoids using deep nesting.

Timothyoung97 commented on src/main/java/command/DoneCommand.java:
```diff
@@ -0,0 +1,33 @@
+public class DoneCommand extends Command {
+
+    private int taskNumber;
+    private final boolean EXIT = false;
```
I like how you make `EXIT` as `private final` and always set to `false` for any command that is not `exit`

Timothyoung97 commented on src/main/java/exception/NullDescription.java:
```diff
@@ -0,0 +1,14 @@
+    public NullDescription(String type) {
+        this.type = type;
+    }
+    @Override
```
I think there should be a space between lines 9 and 10?

KengXIII replied:
Thanks for the spot!

Timothyoung97 reviewed (commented):
LGTM 😀
I like how you extract out commands as abstract classes, it keeps your code relatively easy to read!
Overall, I think it is just the switch cases that require some formatting changes 👍

Timothyoung97 commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,41 @@
+            Task deletedTask = tasks.deleteTask(taskNumber);
+            storage.write(tasks.getTaskList());
+            System.out.printf("\tNoted. I've removed this task:\n" +
+                    "\t%s\n" +
+                    "\tNow you have %d tasks in the list.\n", deletedTask, tasks.size());
```
I feel that maybe you can try using `String.format()` to help tidy up the coding appearance?

JasonC01 reviewed (commented):
The code looks great! But one thing that caught my eyes were the boolean EXIT which kept on appearing at different parts of the code. Maybe its usage can be unified in some way so that EXIT does not have to be repeated. Overall the code really is great!

KengXIII committed:
Change indentation on all switch statements

KengXIII committed:
Rename the output variable to commandSplitBySpace

KengXIII committed:
Fix code convention issues

KengXIII committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

KengXIII committed:
Fixed checkstyle issue in AddCommand

KengXIII committed:
Configure for javafx

KengXIII committed:
Include Javafx for UI

KengXIII committed:
Finish the GUI for duke using fxml

KengXIII committed:
Remove unnecessary imports and fixed coding convention bug in Duke

KengXIII committed:
Add the goodbye method in Ui

KengXIII committed:
Ammend command functions to return String instead of printing

KengXIII committed:
Delete the unused showGoodbye method in Ui

KengXIII committed:
Format the text returned by Duke

KengXIII committed:
Removed tabs from the return String from Commands

KengXIII committed:
Move the exit method into MainWindow class

KengXIII committed:
AddCommand class: newTask is initiated with null

There is no null check for newTask before moving on to add the tasl
into taskList.

An assertion will detect cases where no task happens to be initialised
and replace the null value.

This will verify the correctness of the program flow when Duke runs
and show the error message.

KengXIII committed:
Command class: catch CommandType not properly initialised

CommandType in the parse method may not be initialised if it falls
through all the cases.

The returned value of parse would be a null value.

Assert CommandType to be not equals to true will catch the error cases
and show the error during runtime.

KengXIII committed:
Change the task type from String to enum

KengXIII committed:
Refactor code in AddCommand

KengXIII committed:
Merge pull request #1 from KengXIII/branch-A-Assertions

Insert assert statements into the Command classes

KengXIII committed:
Merge enum tasktype from master

KengXIII committed:
Merge pull request #2 from KengXIII/branch-A-CodeQuality

Refactor code in AddCommand class

KengXIII committed:
Remove time field from deadline and event

KengXIII committed:
Change the datetime input parser format to dd/mm/yyyy

KengXIII committed:
Task Class: Add a new field to store notes for a task.

Task did not support the function to store additional notes
related to task eg. bring a pen.

Adding the notes field allow all tasks to store some relevant info.

KengXIII committed:
Merge pull request #4 from KengXIII/BCD-Extension

Add relevant notes for a task

KengXIII committed:
Fix dialog box to fit according to text length

KengXIII committed:
Remove exit commands that are no longer required for GUI version

KengXIII committed:
Add missing notes compatibility

KengXIII committed:
Write test for deadline and event classes

KengXIII committed:
Add a more extensive catch of invalid formats when adding tasks

KengXIII committed:
Fix the issue where data is not saved when running as jar

KengXIII committed:
Fix date format in toString method of Deadline

KengXIII committed:
Set theme jekyll-theme-slate

KengXIII committed:
Update Readme in docs folder

KengXIII committed:
Merge branch 'master' of https://github.com/KengXIII/ip

KengXIII committed:
Remove unused Ui class

KengXIII committed:
Fix JavaDoc coding convention in Command classes

KengXIII committed:
Fix the string returned by the find method

KengXIII committed:
Add find method guide to ReadMe

KengXIII committed:
Remove arrowhead style code in Storage class

KengXIII committed:
Remove arrowhead style code in Storage class

KengXIII committed:
Fixed all JavaDoc convention violations

KengXIII committed:
Add find to the list of command in ReadMe

KengXIII committed:
Fix Javadoc comment in MainWindow
